### ISSUE TITLE
Uproot attempts to convert NumPy `dtype=object` array directly into Awkward Array

### ISSUE THREAD

**Issue by Superharz** [scikit-hep/uproot5#1101]
Follow up of #1099 

Python version: 3.11.6
Uproot version: 5.2.2rc2.dev13+gc128e4b (installed via pip from the main branch after #1100 had been merged)
zstandard version: 0.22.0 (I had to install it when I first tried to read in the file)
 
Trying to read the `parentLinks` or `childLinks` from a `TruthParticle` in a PhysLite file sometimes results in an error and sometimes works while running on the exact same input file.

The typename of the branch is:
`std::vector<std::vector<ElementLink<DataVector<xAOD::TruthParticle_v1>>>>`

Reading in with numpy works:

```
with uproot.open(path + filename) as f:
    data = f[tree]["TruthBottomAuxDyn.childLinks"].array(library="np")

data
array([<STLVector [] at 0x01a123aa3dd0>, <STLVector [] at 0x01a123bae310>,
       <STLVector [] at 0x01a123bce450>, ...,
       <STLVector [] at 0x01a1204fdb50>, <STLVector [] at 0x01a1204fdb90>,
       <STLVector [] at 0x01a1204fdbd0>], dtype=object)
```

The first none-empty entry contains:
```
[[<ElementLink<DataVector<xAOD::TruthParticle_v1>> (version 1) at 0x01a123ad3550>],
 [<ElementLink<DataVector<xAOD::TruthParticle_v1>> (version 1) at 0x01a123ad26d0>],
 [<ElementLink<DataVector<xAOD::TruthParticle_v1>> (version 1) at 0x01a123ad1b10>,
  <ElementLink<DataVector<xAOD::TruthParticle_v1>> (version 1) at 0x01a123ad3e10>,
  <ElementLink<DataVector<xAOD::TruthParticle_v1>> (version 1) at 0x01a12384bc50>,
  <ElementLink<DataVector<xAOD::TruthParticle_v1>> (version 1) at 0x01a123a806d0>],
 [<ElementLink<DataVector<xAOD::TruthParticle_v1>> (version 1) at 0x01a123a81290>]]
```

The first of these single entries has:
```
{'m_persKey': 779635413,
 'm_persIndex': 3,
 '_typename': 'ElementLink<DataVector<xAOD::TruthParticle_v1>>'}
```

When reading with awkward, 10 out of 20 attempts succeed (I counted) and the other 10 throw an error (again, always reading the exact same file).
The succeeding ones:
```
with uproot.open(path + filename) as f:
    print(f[tree]["TruthBottomAuxDyn.childLinks"].typename)
    data = f[tree]["TruthBottomAuxDyn.childLinks"].array()

data
[[],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 ...,
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 []]
--------------------------------------------------------------------------------
type: 40000 * var * var * struct[{
    m_persKey: uint32,
    m_persIndex: uint32
}, parameters={"__record__": "ElementLink<DataVector<xAOD::TruthParticle_v1>>"}]

```

The failing ones:

```
with uproot.open(path + filename) as f:
    print(f[tree]["TruthBottomAuxDyn.childLinks"].typename)
    data = f[tree]["TruthBottomAuxDyn.childLinks"].array()

data
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[105], line 10
      8 with uproot.open(path + filename) as f:
      9     print(f[tree]["TruthBottomAuxDyn.childLinks"].typename)
---> 10     data = f[tree]["TruthBottomAuxDyn.childLinks"].array()
     12 data

File ~\miniforge3\envs\test\Lib\site-packages\uproot\behaviors\TBranch.py:1815, in TBranch.array(self, interpretation, entry_start, entry_stop, decompression_executor, interpretation_executor, array_cache, library, ak_add_doc)
   1812                 ranges_or_baskets.append((branch, basket_num, range_or_basket))
   1814 interp_options = {"ak_add_doc": ak_add_doc}
-> 1815 _ranges_or_baskets_to_arrays(
   1816     self,
   1817     ranges_or_baskets,
   1818     branchid_interpretation,
   1819     entry_start,
   1820     entry_stop,
   1821     decompression_executor,
   1822     interpretation_executor,
   1823     library,
   1824     arrays,
   1825     False,
   1826     interp_options,
   1827 )
   1829 _fix_asgrouped(
   1830     arrays,
   1831     expression_context,
   (...)
   1835     ak_add_doc,
   1836 )
   1838 if array_cache is not None:

File ~\miniforge3\envs\test\Lib\site-packages\uproot\behaviors\TBranch.py:3142, in _ranges_or_baskets_to_arrays(hasbranches, ranges_or_baskets, branchid_interpretation, entry_start, entry_stop, decompression_executor, interpretation_executor, library, arrays, update_ranges_or_baskets, interp_options)
   3139     pass
   3141 elif isinstance(obj, tuple) and len(obj) == 3:
-> 3142     uproot.source.futures.delayed_raise(*obj)
   3144 else:
   3145     raise AssertionError(obj)

File ~\miniforge3\envs\test\Lib\site-packages\uproot\source\futures.py:38, in delayed_raise(exception_class, exception_value, traceback)
     34 def delayed_raise(exception_class, exception_value, traceback):
     35     """
     36     Raise an exception from a background thread on the main thread.
     37     """
---> 38     raise exception_value.with_traceback(traceback)

File ~\miniforge3\envs\test\Lib\site-packages\uproot\behaviors\TBranch.py:3111, in _ranges_or_baskets_to_arrays.<locals>.basket_to_array(basket)
   3108 basket = None
   3110 if len(basket_arrays) == branchid_num_baskets[branch.cache_key]:
-> 3111     arrays[branch.cache_key] = interpretation.final_array(
   3112         basket_arrays,
   3113         entry_start,
   3114         entry_stop,
   3115         branch.entry_offsets,
   3116         library,
   3117         branch,
   3118         interp_options,
   3119     )
   3120     # no longer needed, save memory
   3121     basket_arrays.clear()

File ~\miniforge3\envs\test\Lib\site-packages\uproot\interpretation\objects.py:475, in AsObjects.final_array(self, basket_arrays, entry_start, entry_stop, entry_offsets, library, branch, options)
    473     output = awkward.concatenate(trimmed, mergebool=False, highlevel=False)
    474 else:
--> 475     output = numpy.concatenate(trimmed)
    477 self.hook_before_library_finalize(
    478     basket_arrays=basket_arrays,
    479     entry_start=entry_start,
   (...)
    484     output=output,
    485 )
    487 output = library.finalize(
    488     output, branch, self, entry_start, entry_stop, options
    489 )

File ~\miniforge3\envs\test\Lib\site-packages\awkward\highlevel.py:1527, in Array.__array_function__(self, func, types, args, kwargs)
   1513 def __array_function__(self, func, types, args, kwargs):
   1514     """
   1515     Intercepts attempts to pass this Array to those NumPy functions other
   1516     than universal functions that have an Awkward equivalent.
   (...)
   1525     See also #__array_ufunc__.
   1526     """
-> 1527     return ak._connect.numpy.array_function(
   1528         func, types, args, kwargs, behavior=self._behavior, attrs=self._attrs
   1529     )

File ~\miniforge3\envs\test\Lib\site-packages\awkward\_connect\numpy.py:102, in array_function(func, types, args, kwargs, behavior, attrs)
    100 function = implemented.get(func)
    101 if function is not None:
--> 102     return function(*args, **kwargs)
    103 # Use NumPy's implementation
    104 else:
    105     all_arguments = chain(args, kwargs.values())

File ~\miniforge3\envs\test\Lib\site-packages\awkward\_connect\numpy.py:142, in implements.<locals>.decorator.<locals>.ensure_valid_args(*args, **kwargs)
    138     names = ", ".join(provided_invalid_names)
    139     raise TypeError(
    140         f"Awkward NEP-18 overload was provided with unsupported argument(s): {names}"
    141     )
--> 142 return function(*args, **kwargs)

File ~\miniforge3\envs\test\Lib\site-packages\awkward\_dispatch.py:62, in named_high_level_function.<locals>.dispatch(*args, **kwargs)
     60 # Failed to find a custom overload, so resume the original function
     61 try:
---> 62     next(gen_or_result)
     63 except StopIteration as err:
     64     return err.value

File ~\miniforge3\envs\test\Lib\site-packages\awkward\operations\ak_concatenate.py:66, in concatenate(arrays, axis, mergebool, highlevel, behavior, attrs)
     63     yield arrays
     65 # Implementation
---> 66 return _impl(arrays, axis, mergebool, highlevel, behavior, attrs)

File ~\miniforge3\envs\test\Lib\site-packages\awkward\operations\ak_concatenate.py:114, in _impl(arrays, axis, mergebool, highlevel, behavior, attrs)
    112 # Now that we're sure `arrays` is not a singular array
    113 with HighLevelContext(behavior=behavior, attrs=attrs) as ctx:
--> 114     content_or_others = ensure_same_backend(
    115         *(
    116             ctx.unwrap(
    117                 x,
    118                 allow_record=axis != 0,
    119                 allow_unknown=False,
    120                 primitive_policy="pass-through",
    121             )
    122             for x in arrays
    123         )
    124     )
    126 contents = [x for x in content_or_others if isinstance(x, ak.contents.Content)]
    127 if len(contents) == 0:

File ~\miniforge3\envs\test\Lib\site-packages\awkward\operations\ak_concatenate.py:116, in <genexpr>(.0)
    112 # Now that we're sure `arrays` is not a singular array
    113 with HighLevelContext(behavior=behavior, attrs=attrs) as ctx:
    114     content_or_others = ensure_same_backend(
    115         *(
--> 116             ctx.unwrap(
    117                 x,
    118                 allow_record=axis != 0,
    119                 allow_unknown=False,
    120                 primitive_policy="pass-through",
    121             )
    122             for x in arrays
    123         )
    124     )
    126 contents = [x for x in content_or_others if isinstance(x, ak.contents.Content)]
    127 if len(contents) == 0:

File ~\miniforge3\envs\test\Lib\site-packages\awkward\_layout.py:146, in HighLevelContext.unwrap(self, obj, allow_record, allow_unknown, none_policy, primitive_policy, string_policy, use_from_iter, regulararray)
    142 from awkward.operations.ak_to_layout import _impl as to_layout_impl
    144 self.update(obj)
--> 146 return to_layout_impl(
    147     obj,
    148     allow_record=allow_record,
    149     allow_unknown=allow_unknown,
    150     none_policy=none_policy,
    151     use_from_iter=use_from_iter,
    152     primitive_policy=primitive_policy,
    153     string_policy=string_policy,
    154     regulararray=regulararray,
    155 )

File ~\miniforge3\envs\test\Lib\site-packages\awkward\operations\ak_to_layout.py:177, in _impl(obj, allow_record, allow_unknown, none_policy, regulararray, use_from_iter, primitive_policy, string_policy)
    175     return obj.snapshot()
    176 elif numpy.is_own_array(obj):
--> 177     promoted_layout = ak.operations.from_numpy(
    178         obj,
    179         regulararray=regulararray,
    180         recordarray=True,
    181         highlevel=False,
    182     )
    183     return _handle_array_like(
    184         obj, promoted_layout, primitive_policy=primitive_policy
    185     )
    186 elif Cupy.is_own_array(obj):

File ~\miniforge3\envs\test\Lib\site-packages\awkward\_dispatch.py:39, in named_high_level_function.<locals>.dispatch(*args, **kwargs)
     35 @wraps(func)
     36 def dispatch(*args, **kwargs):
     37     # NOTE: this decorator assumes that the operation is exposed under `ak.`
     38     with OperationErrorContext(name, args, kwargs):
---> 39         gen_or_result = func(*args, **kwargs)
     40         if isgenerator(gen_or_result):
     41             array_likes = next(gen_or_result)

File ~\miniforge3\envs\test\Lib\site-packages\awkward\operations\ak_from_numpy.py:55, in from_numpy(array, regulararray, recordarray, highlevel, behavior, attrs)
     11 @high_level_function()
     12 def from_numpy(
     13     array,
   (...)
     19     attrs=None,
     20 ):
     21     """
     22     Args:
     23         array (np.ndarray): The NumPy array to convert into an Awkward Array.
   (...)
     52     See also #ak.to_numpy and #ak.from_cupy.
     53     """
     54     return wrap_layout(
---> 55         from_arraylib(array, regulararray, recordarray),
     56         highlevel=highlevel,
     57         behavior=behavior,
     58     )

File ~\miniforge3\envs\test\Lib\site-packages\awkward\_layout.py:347, in from_arraylib(array, regulararray, recordarray)
    344         return ByteMaskedArray(Index8(mask), data, valid_when=False)
    346 if array.dtype == np.dtype("O"):
--> 347     raise TypeError("Awkward Array does not support arrays with object dtypes.")
    349 if isinstance(array, numpy.ma.MaskedArray):
    350     mask = numpy.ma.getmask(array)

TypeError: Awkward Array does not support arrays with object dtypes.

This error occurred while calling

    ak.concatenate(
        [array([<STLVector [] at 0x01a12231aa90>, <STLVector [] at 0x01a11c9b...
    )
```

So it looks like sometimes it falls back to numpy and than fails by attempting to concatenate the STLVector objects.
However, the result is non-deterministic as it sometimes succeeds with creating the awkward array and sometimes it fails.

**Comment by Superharz** [comment]
It looks similar to issue #951
However, my file has been created with a newer ATLAS Athena release: `Athena_24.0.12`

**Comment by jpivarski** [comment]
Could you provide a file? I think this depends on the exact combination of empty `std::vectors` in baskets, which I think is leading to the `trimmed` list having a mix of Awkward Arrays and NumPy `dtype=object` arrays. The latter need to be converted into Awkward Arrays (we have a utility function for this) to fix it, but I want to see it failing to do this in a test-driven way.

**Comment by Superharz** [comment]
Can you access ATLAS internal files?

**Comment by jpivarski** [comment]
> Can you access ATLAS internal files?

Nope. (I can't access CMS internal files, either, and I'm a CMS member.)

But I may have files from #951; it seems to be the same issue.

**Comment by ekourlit** [comment]
We can provide some latest ATLAS PHYSLITE files to non-collaborators. But let me ask @Superharz , do you test on a PHYSLITE file containing real data or Monte Carlo events? Because we see some different behaviour that we try to understand/disentangle...

**Comment by Superharz** [comment]
I am using MC events from the MC23 campaign. 
I am using file `DAOD_PHYSLITE.35010028._000001.pool.root.1` from
`mc23_13p6TeV.601191.PhPy8EG_AZNLO_Ztautau.deriv.DAOD_PHYSLITE.e8514_s4162_r14622_p5855`

**Comment by ivukotic** [comment]
to make things easier I got that file in xcache. Here:
root://xcache.af.uchicago.edu//root://xrootd.echo.stfc.ac.uk:1094/atlas:datadisk/rucio/mc23_13p6TeV/26/c7/DAOD_PHYSLITE.35010028._000001.pool.root.1 /dev/null
